### DISCUSSION TITLE
Issue binding data to model?

### DISCUSSION THREAD

**Discussion by dgomez-listrak** [MarkPflug/Sylvan#85] [Q&A] [answered]
I'm trying to follow the guide below to bind the CSV data directly to my model, and while it's generating the expected number of items in the return value, no actual data is being bound.  What am I missing?

In https://github.com/MarkPflug/Sylvan/blob/e4df44df68b4489a4cef1d249528a67ed6bf1014/docs/Csv/Examples.md

My code snippet:

 ```
public static IEnumerable<receiptModel> GetRecords(string csvFile)
        {
            using var csv = CsvDataReader.Create(csvFile);
            var binder = DataBinder.Create<receiptModel>(csv);            
            while (csv.Read())
            {
                var record = new receiptModel();
                binder.Bind(csv, record); // had to at first param; example was missing
                yield return record;
            }
        }
```

**Comment by MarkPflug** [comment]
Can you share the definition of the `receiptModel`? Is it a struct, by any chance? Binding to structs isn't currently supported.
What version of the Sylvan.Data package are you using?

**Reply by dgomez-listrak** [reply]
`    public class receiptModel
    {
        public string storeId;
        public string registerId;
        public string transNo;
        public string lineItem;
        public string lineType;
        public string receiptLine;
    }
`

Using version 0.1.0-b0024.

**Comment by MarkPflug** [comment]
Change the fields to properties and it should work.

```
public class receiptModel { 
public string storeId {get;set;};
public string registerId{get;set;} 
public string transNo{get;set;} 
public string lineItem{get;set;}
public string lineType{get;set;}
public string receiptLine{get;set;}
 }
```

The databinder needs some work, so I'm taking notes about some potential improvements.

**Comment by dgomez-listrak** [comment]
Had this error thrown when I did that on this line:             var binder = DataBinder.Create<receiptModel>(csv);

Unhandled exception. Sylvan.Data.DataBinderException: Exception of type 'Sylvan.Data.DataBinderException' was thrown.
   at Sylvan.Data.CompiledDataBinder`1..ctor(DataBinderOptions opts, ReadOnlyCollection`1 physicalSchema, ReadOnlyCollection`1 logicalSchema)
   at Sylvan.Data.CompiledDataBinder`1..ctor(DataBinderOptions opts, ReadOnlyCollection`1 physicalSchema)
   at Sylvan.Data.DataBinder.Create[T](IDataReader dr, DataBinderOptions opts)

This is my code ( I might be fundamentally missing something here that's not standing out)

```
using Sylvan.Data;
using Sylvan.Data.Csv;
using System.Collections.Generic;
using System.Data;
using System.Threading.Tasks;
namespace csvprocessor
{
    public class receiptModel
    {
        public int storeId { get; set; }
        public int registerId { get; set; }
        public int transNo { get; set; }
        public string lineItem { get; set; }
        public string lineType { get; set; }
        public string receiptLine { get; set; }
        
    }
    public class Program
    {
        static async Task Main(string[] args)
        {         
            var csvRecords = GetRecords("C:\\Dropbox\\work\\csvfile.csv");
            foreach (var record in csvRecords)            
                System.Console.WriteLine(record.storeId);
                          
            System.Console.WriteLine("fin");
        }


        public static IEnumerable<receiptModel> GetRecords(string csvFile)
        {

            using var csv = CsvDataReader.Create(csvFile, new CsvDataReaderOptions
            {
                HasHeaders = false,                
            });
            var binder = DataBinder.Create<receiptModel>(csv);

            while (csv.Read())
            {
                var record = new receiptModel();
                binder.Bind(csv, record);
                yield return record;
            }
        }        
    }
}
```

**Comment by MarkPflug** [comment, marked as answer]
It looks like the issue is that your csv file doesn't have headers (`HasHeaders = false`). The binder doesn't know what to do with the data, since it isn't explicitly labeled. You can fix this by defining the column schema explicitly.

```
public static IEnumerable<receiptModel> GetRecords(string csvFile)
        {
            var schema = Schema.Parse("storeId,registerId,transNo,lineItem,lineType,receiptLine");

            var options = new CsvDataReaderOptions
            {
                HasHeaders = false,
                Schema = new CsvSchema(schema),
            };

            using var csv = CsvDataReader.Create(csvFile, options);
            var binder = DataBinder.Create<receiptModel>(csv);

            while (csv.Read())
            {
                var record = new receiptModel();
                binder.Bind(csv, record);
                yield return record;
            }
        }
```

**Reply by dgomez-listrak** [reply]
Ahah!  That got it.  Thanks for the help!
